src/storage/src/memtable/btree.rs

```diff
@@ -196,25 +204,53 @@ impl BTreeIterator {
         let key_columns = rows_to_vectors(
             key_data_types,
             self.adapter.source_key_needed(),
+            &row_selection,
             keys.as_slice(),
         );
         let field_columns = rows_to_vectors(
             value_data_types,
             self.adapter.source_value_needed(),
+            &row_selection,
             values.as_slice(),
         );
 
+        let sequences = arrow::compute::take(&sequences.to_arrow_array(), &row_selection, None)
+            .context(error::SortArraysSnafu)?;
+        let op_type = arrow::compute::take(&op_types.to_arrow_array(), &row_selection, None)
+            .context(error::SortArraysSnafu)?;
+
         let batch = self.adapter.batch_from_parts(
             key_columns,
             field_columns,
-            Arc::new(sequences),
-            Arc::new(op_types),
+            Helper::try_into_vector(sequences)
```

Comment:
issue: If all rows are filtered then we might return an empty vector. I remember that batch_from_parts doesn't expect empty vectors.